android/src/main/new_arch/react/renderer/components/RNLiveMarkdownSpec/MarkdownTextInputDecoratorShadowNode.cpp

```diff
@@ -0,0 +1,122 @@
+#include <react/renderer/core/ComponentDescriptor.h>
+#include <yoga/Yoga.h>
+
+namespace facebook {
+namespace react {
+
+extern const char MarkdownTextInputDecoratorViewComponentName[] =
+    "MarkdownTextInputDecoratorView";
+
+void MarkdownTextInputDecoratorShadowNode::initialize() {
+  ShadowNode::traits_.unset(ShadowNodeTraits::ForceFlattenView);
+}
+
+void MarkdownTextInputDecoratorShadowNode::createCustomContextContainer() {
+  static auto customUIManagerClass = jni::findClassStatic(
+      "com/expensify/livemarkdown/CustomFabricUIManager");
+  static auto createCustomUIManager =
+      customUIManagerClass
+          ->getStaticMethod<JFabricUIManager::javaobject(
+              JFabricUIManager::javaobject, ReadableMap::javaobject, int)>(
+              "create");
+
+  const auto markdownStyle =
+      this->getProps()->rawProps["markdownStyle"];
+  const auto currentParserId =
+      this->getProps()->rawProps["parserId"].asInt();
+
+  auto const decoratorPropsRNM =
+      ReadableNativeMap::newObjectCxxArgs(markdownStyle);
+  auto const decoratorPropsRM =
+      jni::make_local(reinterpret_cast<ReadableMap::javaobject>(
+                          decoratorPropsRNM.get()));
+
+  const JFabricUIManager::javaobject &fabricUIManager =
+      this->getContextContainer()->at<JFabricUIManager::javaobject>("FabricUIManager");
+
+  const auto customUIManager = jni::make_global(createCustomUIManager(
+      customUIManagerClass, fabricUIManager,
+      decoratorPropsRM.get(), currentParserId));
+  const ContextContainer::Shared contextContainer =
+      std::make_shared<ContextContainer const>();
+  contextContainer->insert("FabricUIManager", customUIManager);
+
+  customContextContainer_ = contextContainer;
+  previousMarkdownStyle_ = markdownStyle;
+  previousParserId_ = currentParserId;
+}
+
+void MarkdownTextInputDecoratorShadowNode::tryUpdateCustomContextContainer() {
+  const auto markdownStyle =
+      this->getProps()->rawProps["markdownStyle"];
+  const auto currentParserId =
+      this->getProps()->rawProps["parserId"].asInt();
+
+  if (currentParserId != previousParserId_ || markdownStyle != previousMarkdownStyle_) {
+    createCustomContextContainer();
+  }
+}
+
+void MarkdownTextInputDecoratorShadowNode::adoptChildren() {
+  const auto &children = getChildren();
+  if (children.empty()) {
+    return;
+  }
+  react_native_assert(
+      children.size() == 1 &&
+      "MarkdownTextInputDecoratorView received more than one child");
+  react_native_assert(
+      std::dynamic_pointer_cast<const AndroidTextInputShadowNode>(children.at(0)) &&
+      "MarkdownTextInputDecoratorView received child other than a TextInput");
+
+  const auto child = std::dynamic_pointer_cast<const AndroidTextInputShadowNode>(
+      children.at(0));
+  const auto mutableChild = std::const_pointer_cast<AndroidTextInputShadowNode>(child);
+  mutableChild->setTextLayoutManager(
+      std::make_shared<TextLayoutManager>(customContextContainer_));
```

Comment:
Perhaps we should start memoizing an instance of `TextLayoutManager` at some point.